File: crates/herkos-tests/data/c/c_e2e_loops.c
// Loop-heavy algorithms: power, collatz, primality, divisor sum

// Exponentiation by repeated multiplication
int power(int base, int exp) {
    int result = 1;
    while (exp > 0) {
        result *= base;
        exp--;
    }
    return result;
}

// Count steps in the Collatz sequence until reaching 1
int collatz_steps(int n) {
    int steps = 0;
    while (n > 1) {
        if (n & 1) {
            n = 3 * n + 1;
        } else {
            n = n >> 1;
        }
        steps++;
    }
    return steps;
}

// Trial division primality test. Returns 1 if prime, 0 if not.
// Avoids ternary / select by using a flag variable.
int is_prime(int n) {
    if (n < 2) { return 0; }
    int i = 2;
    while (i * i <= n) {
        if (n % i == 0) { return 0; }
        i++;
    }
    return 1;
}

// Count the number of primes <= n (simple sieve-free approach)
int count_primes(int n) {
    int count = 0;
    int i = 2;
    while (i <= n) {
        count += is_prime(i);
        i++;
    }
    return count;
}

// Sum of proper divisors of n (excluding n itself)
int sum_of_divisors(int n) {
    int sum = 0;
    int i = 1;
    while (i * i <= n) {
        if (n % i == 0) {
            sum += i;
            if (i != 1 && i != n / i) {
                sum += n / i;
            }
        }
        i++;
    }
    return sum;
}

// Check if n is a perfect number (sum of divisors == n)
int is_perfect(int n) {
    if (n < 2) { return 0; }
    int s = sum_of_divisors(n);
    // Avoid ternary: use subtraction trick
    // Returns 1 if s == n, 0 otherwise
    int diff = s - n;
    // (diff == 0) => !(diff | (-diff)) >> 31 & 1
    // Simpler: just use equality via subtraction and or
    return (diff == 0) & 1;
}

// Digital root: repeatedly sum digits until single digit
int digital_root(int n) {
    while (n >= 10) {
        int sum = 0;
        while (n > 0) {
            sum += n % 10;
            n /= 10;
        }
        n = sum;
    }
    return n;
}

File: crates/herkos-tests/tests/c_e2e_loops.rs
//! End-to-end tests: C → Wasm → Rust (loop-heavy algorithms).

use herkos_tests::c_e2e_loops;

fn new_module() -> c_e2e_loops::WasmModule {
    c_e2e_loops::new().expect("module instantiation should succeed")
}

// ── Power ──

#[test]
fn test_power() {
    let mut m = new_module();
    assert_eq!(m.power(2, 0).unwrap(), 1);
    assert_eq!(m.power(2, 1).unwrap(), 2);
    assert_eq!(m.power(2, 10).unwrap(), 1024);
    assert_eq!(m.power(3, 5).unwrap(), 243);
    assert_eq!(m.power(1, 100).unwrap(), 1);
    assert_eq!(m.power(0, 5).unwrap(), 0);
    assert_eq!(m.power(-1, 3).unwrap(), -1);
    assert_eq!(m.power(-1, 4).unwrap(), 1);
}

// ── Collatz ──

#[test]
fn test_collatz_steps() {
    let mut m = new_module();
    assert_eq!(m.collatz_steps(1).unwrap(), 0);
    assert_eq!(m.collatz_steps(2).unwrap(), 1);
    assert_eq!(m.collatz_steps(3).unwrap(), 7);
    assert_eq!(m.collatz_steps(6).unwrap(), 8);
    assert_eq!(m.collatz_steps(27).unwrap(), 111);
}

// ── Primality ──

#[test]
fn test_is_prime() {
    let mut m = new_module();
    assert_eq!(m.is_prime(0).unwrap(), 0);
    assert_eq!(m.is_prime(1).unwrap(), 0);
    assert_eq!(m.is_prime(2).unwrap(), 1);
    assert_eq!(m.is_prime(3).unwrap(), 1);
    assert_eq!(m.is_prime(4).unwrap(), 0);
    assert_eq!(m.is_prime(5).unwrap(), 1);
    assert_eq!(m.is_prime(17).unwrap(), 1);
    assert_eq!(m.is_prime(18).unwrap(), 0);
    assert_eq!(m.is_prime(97).unwrap(), 1);
    assert_eq!(m.is_prime(100).unwrap(), 0);
}

#[test]
fn test_count_primes() {
    let mut m = new_module();
    assert_eq!(m.count_primes(1).unwrap(), 0);
    assert_eq!(m.count_primes(10).unwrap(), 4); // 2,3,5,7
    assert_eq!(m.count_primes(20).unwrap(), 8); // 2,3,5,7,11,13,17,19
    assert_eq!(m.count_primes(100).unwrap(), 25);
}

// ── Divisors ──

#[test]
fn test_sum_of_divisors() {
    let mut m = new_module();
    // sum_of_divisors returns sum of proper divisors (excluding n)
    assert_eq!(m.sum_of_divisors(1).unwrap(), 1); // just 1
    assert_eq!(m.sum_of_divisors(6).unwrap(), 6); // 1+2+3
    assert_eq!(m.sum_of_divisors(28).unwrap(), 28); // 1+2+4+7+14
    assert_eq!(m.sum_of_divisors(12).unwrap(), 16); // 1+2+3+4+6
}

#[test]
fn test_is_perfect() {
    let mut m = new_module();
    assert_eq!(m.is_perfect(6).unwrap(), 1);
    assert_eq!(m.is_perfect(28).unwrap(), 1);
    assert_eq!(m.is_perfect(496).unwrap(), 1);
    assert_eq!(m.is_perfect(10).unwrap(), 0);
    assert_eq!(m.is_perfect(1).unwrap(), 0);
}

// ── Digital root ──

#[test]
fn test_digital_root() {
    let mut m = new_module();
    assert_eq!(m.digital_root(0).unwrap(), 0);
    assert_eq!(m.digital_root(5).unwrap(), 5);
    assert_eq!(m.digital_root(9).unwrap(), 9);
    assert_eq!(m.digital_root(10).unwrap(), 1);
    assert_eq!(m.digital_root(99).unwrap(), 9);
    assert_eq!(m.digital_root(123).unwrap(), 6); // 1+2+3=6
    assert_eq!(m.digital_root(9999).unwrap(), 9);
}
